Test to verify mirror pool replica size changes when mirroring I/O are in progress

sunilangadi2 opened the pull request:
# Description
Automated test case - https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitem?id=CEPH-9518
Fixes:
RHCEPHQE-7885

**Test Case Flow:**
      1. Create a pool on both clusters.
      2. Create an Image on cluster 1 and create a clone of the image in same pool.
      3. Configure mirroring (peer bootstrap) between two clusters.
      4. Enable pool mode journal based mirroring on the pool respectively.
      5. Start running IOs on the primary image.
      6. While IOs is running decrease the replica count of mirror pool,
      (if rep count is 3 decrease to 2).
      7. IO's should not stop and cluster health status should be healthy
      8. While IOs is running increase the replica count of mirror pool,
      (if rep count is 2 increase to 3).
      9. IO's should not stop and cluster health status should be healthy
  
Test Results:
[Passed] http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-LXJA0J
  
  Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
  
  <details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

rkhambad commented on tests/rbd_mirror/test_rbd_mirror_replica_count.py:
```diff
@@ -0,0 +1,123 @@
+                    results,
+                    cmd=f"ceph osd pool set {poolname} size {new_size}",
+                )
+
```
IMO, we should cross check the value we are setting for replica size.

VasishtaShastry reviewed (commented):
I see that TC mentions that below things are expected to be verified
"Rebalancing should start"
"Cluster health should become healthy"

IMO, this should at least be check here or at least in tier-3 scenarios (add to some TCs if there are not some specifics.)
Please let me know your thoughts

sunilangadi2 commented:
> I see that TC mentions that below things are expected to be verified "Rebalancing should start" "Cluster health should become healthy"
> 
> IMO, this should at least be check here or at least in tier-3 scenarios (add to some TCs if there are not some specifics.) Please let me know your thoughts

Thanks @VasishtaShastry for the review,
in the logs if you see http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-LXJA0J/Test_to_change_of_mirror_pool_replica_size_0.log
ceph health status `io` part has `recovery` due to change of pool replica size
Would it be fine to validate by having some values in `recovery` ?

Due to replica size change PG get's degrade so it takes more time all PG's to get back `active and clean`
hence getting Health_OK status we need to wait for more time it might be days as well
Hence i validated using `HEALTH_ERR` status if by any of these test actions bring a health `error state` we can fail it.
Please let me know your thoughts or do you have any other way of validating it?

sunilangadi2 commented:
> > I see that TC mentions that below things are expected to be verified "Rebalancing should start" "Cluster health should become healthy"
> > IMO, this should at least be check here or at least in tier-3 scenarios (add to some TCs if there are not some specifics.) Please let me know your thoughts
> 
> Thanks @VasishtaShastry for the review, in the logs if you see http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-LXJA0J/Test_to_change_of_mirror_pool_replica_size_0.log ceph health status `io` part has `recovery` due to change of pool replica size Would it be fine to validate by having some values in `recovery` ?
> 
> Due to replica size change PG get's degrade so it takes more time all PG's to get back `active and clean` hence getting Health_OK status we need to wait for more time it might be days as well Hence i validated using `HEALTH_ERR` status if by any of these test actions bring a health `error state` we can fail it. Please let me know your thoughts or do you have any other way of validating it?

Discussed with @VasishtaShastry in PR review meeting 
Plan is to add Data consistency check to make sure recovery of IO should complete and data on mirror site should reflect same
working on implementing those suggestions.
Thanks @VasishtaShastry for the suggestion.

sunilangadi2 commented on tests/rbd_mirror/test_rbd_mirror_replica_count.py:
```diff
@@ -0,0 +1,123 @@
+
+    Test Case Flow:
+        1. Create a pool on both clusters.
+        2. Create an Image on cluster 1 and create a clone of the image in same pool.
```
Removed, thanks for noticing.

sunilangadi2 commented:
@rkhambad @VasishtaShastry Thanks for the review,
Addressed review comments
Test results for the updated code
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-5T6ZDU
@VasishtaShastry after adding `check_data` consistency test for io `1G` it took more than 3hours hence i reduced `io` to `200M`

rkhambad reviewed (approved):
LGTM

manasagowri reviewed (approved):
Group review done. Changes LGTM
